internal/controller/cull.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	amaltheadevv1alpha1 "github.com/SwissDataScienceCenter/amalthea/api/v1alpha1"
+)
+
+func customResourceNeedsDeletion(amaltheasession *amaltheadevv1alpha1.AmaltheaSession) bool {
+	culling := amaltheasession.Spec.Culling
+	status := amaltheasession.Status
+	now := time.Now()
+	hibernatedDuration := now.Sub(status.HibernatedSince.Time)
+	return status.State == amaltheadevv1alpha1.Hibernated &&
+		hibernatedDuration > culling.MaxHibernatedDuration.Duration
+}
+
+func updateStatefulSetReplicas(ctx context.Context, r *AmaltheaSessionReconciler, amaltheasession *amaltheadevv1alpha1.AmaltheaSession) error {
```

Comment:
Following discussion the name of the function is not correct anymore:

```suggestion
func updateHibernationState(ctx context.Context, r *AmaltheaSessionReconciler, amaltheasession *amaltheadevv1alpha1.AmaltheaSession) error {
```